source/common/http/utility.h

```diff
@@ -233,6 +233,106 @@ void transformUpgradeRequestFromH2toH1(HeaderMap& headers);
  */
 void transformUpgradeResponseFromH2toH1(HeaderMap& headers, absl::string_view upgrade);
 
+/**
+ * The non template implementation of resolvePerFilterConfig. see
+ * resolvePerFilterConfig for docs.
+ */
+const Router::RouteSpecificFilterConfig*
+resolvePerFilterConfigGeneric(const std::string& filter_name,
+                              const Router::RouteConstSharedPtr& route);
+
+/**
+ * Retreives the route specific config. Route specific config can be in a few
+ * places, that are checked in order. The first config found is returned. The
+ * order is:
+ * - the routeEntry() (for config that's applied on weighted clusters)
+ * - the route
+ * - and finally from the virtual host object (routeEntry()->virtualhost()).
+ *
+ * To use, simply:
+ *
+ *     const auto* config =
+ *         Utility::resolvePerFilterConfig<ConcreteType>(FILTER_NAME, stream_callbacks_.route());
+ *
+ * See notes about config's lifetime below.
+ *
+ * @param filter_name The name of the filter who's route config should be
+ * fetched.
+ * @param route The route to check for route configs. nullptr routes will
+ * result in nullptr being returned.
+ *
+ * @return The route config if found. nullptr if not found. The returned
+ * pointer's lifetime is the same as the route parameter.
+ */
+template <class ConfigType>
+const ConfigType* resolvePerFilterConfig(const std::string& filter_name,
```

Comment:
Suggest naming these `resolveMostSpecificPerFilterConfig` to make it clearer what is going on. TBH, it seems you should only actually need a `fold` operation here. If you were folding over the chain, you would be building up a single `ExtAuthzPerRoute` message, with different merge vs. override semantics for each field. This would be a bit more elegant.

Reply:
If I Understand Correctly, the reason not to use fold here is to avoid creating copies of the config when not needed.
i.e. if you know you have no need for merging, there's no need to fold and create a copy, so that one can use the copy that's already cached in the route.